### DISCUSSION TITLE
Manifests should install dependencies earlier in build process, before the `CustomBuild` target

### DISCUSSION THREAD

**Discussion by aggieNick02** [microsoft/vcpkg#21601] [Ideas]
**Describe the bug**
Manifests should install dependencies before the `CustomBuild` target because their failure to do so means custom build tool items in a project cannot reliably use `vcpkg` installed dependencies.

This bug/issue is motivated by tools such as `grpc` and `protobuf`, which provide code generation tools to generate `.h`,`.cpp`.,`.py` files and more. A typical way to use them is to add a `.proto` file to your project and provide a custom build action that invokes `protoc.exe` or one of the `grpc` tools, generating files that are also part of the project.

`grpc` and `protobuf` are both ports in `vcpkg`, so `vcpkg` is a natural way to obtain them. Then the custom build action can just point to their tools in the `vcpkg` installed directory, e.g., `"$(_ZVcpkgCurrentInstalledDir)tools\protobuf\protoc.exe"`

The problem comes when using manifest mode. The targets in `vcpkg.targets` are specified with `BeforeTargets=ClCompile`, which means that the manifest dependencies are installed *after* any files with custom build tools specified are processed. That is, in a clean setup where `vcpkg` has not installed anything, a `.proto` file will fail to be processed by the `protobuf` or `grpc` tools that are specified, because those tools do not have a chance to be installed until after the `.proto` file is compiled

Simply changing `VcpkgInstallManifestDependencies` and `VcpkgTripletSelection` in `vcpkg.targets` to have ~`BeforeTargets="CustomBuild"`~ `BeforeTargets="CustomBuild;ClCompile"`addresses the issue. Note: I updated from the strikethrough to include both `ClCompile` and `CustomBuild`, since not specifying `ClCompile` breaks the most common usage :-P . But even this isn't great, as it doesn't cover other build actions, e.g., `CudaBuild` as hit by @dahubley in https://github.com/microsoft/vcpkg/issues/20884 .

Is there any reason to not make this change in `vcpkg.targets`? I'm happy to submit the PR, but I'm not an expert in msbuild and the targets, so I'm not 100% sure that won't have any side effects.

**Environment**
- OS: Windows
- Compiler: Visual Studio 2019

**To Reproduce**
1. .Get `vcpkg` and run `.\bootstrap_vcpkg.bat`, and then `.\vcpkg.exe integrate install`
2. Create a new console app project and solution.
3. Set the new project to use manifests in the vcpkg project properties.
4. Create a vcpkg.json manifest next to the solution with the contents:
```
{
    "$schema": "https://raw.githubusercontent.com/microsoft/vcpkg/master/scripts/vcpkg.schema.json",
    "name": "sample-manifest",
    "version": "0.1",
    "dependencies": [
        {
            "name": "presentmon",
            "features": ["tools"]
        }
    ]
}
```
3. Add a text file to your project. Right click on the text file and choose properties. Change the item type to "Custom Build Tool." Hit ok.
4. Right click on the text file and choose to compile. When it completes, notice that there is not yet a `vcpkg_installed\x86-windows\...` hierarchy next to `vcpkg.json` or any mention in the compile output about `vcpkg` installing dependencies. This is the bug/issue - because custom build tool items don't cause `vcpkg` dependencies to install, they can't reliably use them.
5. Right click on the console application cpp file and choose to compile. Notice the vcpkg manifest dependencies are now installed.

**Expected behavior**
Install manifest dependencies before any custom build items run

**Failure logs** 
N/A

**Additional context**
N/A

**Comment by PhoebeHui** [comment]
@vicroms, could you help double confirm?

**Comment by aggieNick02** [comment]
A little bit related if you have thoughts on this as well: https://github.com/microsoft/vcpkg/discussions/20846 (It would be really nice if `bootstrap_vcpkg.bat` was lightweight enough that it did nothing when vcpkg.exe was up to date, so that it could always be called, either manually or by the `vcpkg.targets` manifest installation step)

**Comment by ghost** [comment]
I think this is the same issue as #20884.

**Comment by aggieNick02** [comment]
Investigating the bug filed by @dahubley made me realize my initial proposed fix is no good. I'm updating my report above with more details.

Even after the issue with my original fix is corrected, using `BeforeTargets` to do the `vcpkg` targets seems problematic because it has to be updated for each kind of build action, e.g., `ClCompile`, `CustomBuild`, `CudaBuild`. We can add `CustomBuild` and `CudaBuild`, but this doesn't cover everything.

**Comment by yog-muskrat** [comment]
I've found that you can make vcpkg run before module scanning by setting the dependency via the SetModuleDependenciesDependsOn variable in your project. The SetModuleDependencies target should then respect it.
E.g.:
```
<SetModuleDependenciesDependsOn>$(SetModuleDependenciesDependsOn);VcpkgInstallManifestDependencies</SetModuleDependenciesDependsOn>
```
However, I'm not sure if this is a public feature, as there is no mention of it anywhere.
